caom2pipe/client_composable.py

```diff
@@ -0,0 +1,665 @@
+#  hope that it will be useful,         dans l’espoir qu’il vous
+#  but WITHOUT ANY WARRANTY;            sera utile, mais SANS AUCUNE
+#  without even the implied             GARANTIE : sans même la garantie
+#  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+#  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+#  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+#  General Public License for           Générale Publique GNU Affero
+#  more details.                        pour plus de détails.
+#
+#  You should have received             Vous devriez avoir reçu une
+#  a copy of the GNU Affero             copie de la Licence Générale
+#  General Public License along         Publique GNU Affero avec
+#  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+#  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+#                                       <http://www.gnu.org/licenses/>.
+#
+#  : 4 $
+#
+# ***********************************************************************
+#
+
+"""
+This module contains classes and functions that have to do with managing
+and using long-term connections to CADC services.
+
+Leave out:
+- get_artifact_metadata_client, since it mostly CAOM2 dependent, and the
+  only client-based part is get_cadc_meta_client, which is already here
+- should Metrics be here? Maybe.
+"""
+
+import logging
+import os
+import traceback
+
+from dataclasses import dataclass
+from datetime import datetime
+from io import BytesIO, StringIO
+from sys import getsizeof
+
+from astropy.table import Table
+
+from cadcdata import CadcDataClient
+from cadctap import CadcTapClient
+from cadcutils import net, exceptions
+from caom2pipe import manage_composable as mc
+from caom2repo import CAOM2RepoClient
+from vos import Client
+
+__all__ = [
+    'client_get',
+    'client_put',
+    'client_put_fqn',
+    'ClientCollection',
+    'current',
+    'data_get',
+    'data_put_fqn',
+    'declare_client',
+    'define_subject',
+    'get_cadc_headers_client',
+    'get_cadc_meta_client',
+    'get_cadc_meta_client_v',
+    'look_pull_and_put',
+    'look_pull_and_put_v',
+    'query_tap_client',
+    'repo_create',
+    'repo_delete',
+    'repo_get',
+    'repo_update',
+]
+
+
+class ClientCollection(object):
+    """
+    This class initializes and provides accessors to known HTTP clients
+    for CADC services.
+
+    This is time-dependent cohesiveness. The clients should all be declared
+    at approximately the same time, will live as long as the pipeline
+    instance, and require the same AAI credential information, even though
+    there is one per not-related-at-all CADC service.
+
+    Eventually session retries, etc, might be configurable, but that need has
+    not yet been demonstrated.
+    """
+
+    def __init__(self, config):
+        self._metadata_client = None
+        self._data_client = None
```

Comment:
General comment. If this needs to support both `data` and `storage` inventory I would use 4 methods (get, put, info, remove) that would call the right client to perform this operations. It would be transparent to the rest of the code what client it's being used.

Reply:
I wish I had thought of this sooner, because it's a great idea.

Unfortunately when I started this change, the prospective client was going to be `vos.Client`, and so I thought I knew what those calls were going to look like, and so I went for specific, instead of general. If you think it's worthwhile to create this class and modify the all the code locations where this class may be of use (including caom2tools.caom2utils.fits2caom2.py), I will.